src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs

```diff
@@ -360,7 +360,17 @@ public override string GetContentHashForSignedPackage(CancellationToken token)
             {
                 return SignedPackageArchiveUtility.GetPackageContentHash(reader);
             }
+        }
+
+        public override string GetContentHash(CancellationToken token)
+        {
+            var contentHash = GetContentHashForSignedPackage(token);
+            if (contentHash == null)
+            {
+                contentHash =  Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(ZipReadStream));
```

Comment:
I *think* this is incorrect.  If `GetContentHashForSignedPackage(…)` returns `null`, it means `ZipArchive` did a bunch of reads on the stream.  The stream's current position is indeterminate; it's wherever `ZipArchive` left off, which may be what you expect or may not; however, it's not part of `ZipArchive`'s contract to leave a stream's current position wherever you expect.  To be safe, you should explicitly seek to the start of the stream using one of the following statements:

    ZipReadStream.Position = 0;
    ZipReadStream.Seek(offset: 0, origin: SeekOrigin.Begin);

Otherwise, you may generate a hash based on an incomplete read of the ZIP stream.  That would be really bad.

Also, format document while you're in here to remove the extra whitespace.